subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/AbstractListChildMap.java

```diff
@@ -122,24 +125,46 @@ protected int findChildIndexWithCommonPrefix(VfsRelativePath targetPath, CaseSen
     }
 
     protected ChildMap<T> withNewChild(int insertBefore, String path, T newChild) {
-        List<Entry<T>> newChildren = new ArrayList<>(entries);
-        newChildren.add(insertBefore, new Entry<>(path, newChild));
+        Entry<T>[] newChildren = createEntryArray(entries.length + 1);
+        if (insertBefore > 0) {
+            System.arraycopy(entries, 0, newChildren, 0, insertBefore);
+        }
+        newChildren[insertBefore] = new Entry<>(path, newChild);
+        if (insertBefore < entries.length) {
+            System.arraycopy(entries, insertBefore, newChildren, insertBefore + 1, entries.length - insertBefore);
+        }
```

Comment:
Is replacing this `ArrayList` with an array really worth this additional complexity? It definitely makes the code much less readable. It would be good to prove that this particular step adds enough bang to be worth it.

Reply:
Replacing the `ArrayList`s with arrays actually saves the biggest amount of garbage in this change, especially for the insert new element use-case. If we use `ArrayList`s this is what happens:
- `new ArrayList(Collection)` converts the collection to an array.
- `ArrayList.add(int,T)` creates another new array and copies over the data

So we have an extra array created when inserting an element. I agree that the garbage created should be collected immediately. It still seems to show up quite a lot in the allocation flames.

Reply:
I might be missing something, but why not do something like `new ArrayList<>(entries.size() + 1)` and then add things in the right places? Essentially doing the same as we do with `System.arraycopy()` but with `ArrayList` instead.

If we want to keep the array-based code, I think we should make sure there's as little low-level code as possible. Extracting the `arraycopy()`-related logic into a single method would make sense if possible.

Reply:
I tried that again and compared. This is what I would do with array lists:
```
    protected ChildMap<T> withNewChild(int insertBefore, String path, T newChild) {
        List<Entry<T>> newChildren = new ArrayList<>(entries.size() + 1);
        newChildren.addAll(entries);
        newChildren.add(insertBefore, new Entry<>(path, newChild));
        return ChildMapFactory.childMapFromSorted(newChildren);
    }
```

This avoids one full allocation and uses `arraycopy` in the call to `addAll`. The problem here is that `addAll` converts the collection passed in to an array by using `toArray()`, so it causes another allocation which shows up in the flame graphs. This means by the code above I get rid of one array allocation, though the other one still remains.